File: Javascript/p005-ReLU-Activation.js
/*
Creates a dense layer of neurons with a ReLU activation function, and feeds forward inputs through them.
Associated YT tutorial: https://www.youtu.be/gmjzbpSVY1A
*/

// get the shape of a nested-array (assuming it is 'rectangular', but in many more or fewer dimensions)
function shape(a) {
	const res = [];
	let temp = a;
	while (Array.isArray(temp)) {
		res.push(temp.length);
		temp = temp[0];
	}
	return res;
}

// functions roughly as np.dot(a, b); (it works fine for this context)
function dot(a, b) {
	const res = [];
	const shapeA = shape(a);
	const shapeB = shape(b);
	if (shapeA.length === 1 && shapeB.length === 1) {
		let res = 0;
		for (let i = 0; i < shapeA[0]; i++) {
			res += a[i]*b[i];
		}
		return res;
	} else if (shapeA.length === 2 && shapeB.length === 1) {
		for (let i = 0; i < a.length; i++) {
			let dot = 0;
			for (let j = 0; j < b.length; j++) {
				dot += a[i][j]*b[j];
			}
			res.push(dot);
		}
	} else if (shapeA.length === 2 && shapeB.length === 2 && shapeA[1] === shapeB[0]) {
		for (let i = 0; i < shapeA[0]; i++) {
			const _dots = [];
			for (let j = 0; j < shapeB[1]; j++) {
				let _dot = 0;
				for (let k = 0; k < shapeB[0]; k++) {
					_dot += a[i][k]*b[k][j];
				}
				_dots.push(_dot);
			}
			res.push(_dots);
		}
	}
	return res;
}

// assumes you are adding b on to a, where b can be a number or an array
function add(a, b) {
	const res = [];
	if (Array.isArray(a) && typeof b === "number") {
		for (let i = 0; i < a.length; i++) {
			res.append(a[i] + b);
		}
		return res;
	} else if (Array.isArray(a) && Array.isArray(b)) {
		const shapeA = shape(a);
		const shapeB = shape(b);
		if (shapeA.length === 1 && shapeB.length === 1) {
			for (let i = 0; i < a.length; i++) {
				res.push(a[i] + b[i]);
			}
		} else if (shapeA.length === 2 && shapeB.length === 1 &&
			shapeA[1] === shapeB[0]) {
			for (let i = 0; i < shapeA[0]; i++) {
				const row = [];
				for (let j = 0; j < shapeA[1]; j++) {
					row.push(a[i][j] + b[j]);
				}
				res.push(row);
			}
		}
	}
	return res;
}

// replaces np.array().T
function transpose(a) {
	const res = [];
	const _shape = shape(a);
	if (_shape.length === 2) {
		for (let i = 0; i < _shape[1]; i++) {
			const row = [];
			for (let j = 0; j < _shape[0]; j++) {
				row.push(a[j][i]);
			}
			res.push(row);
		}
	}
	return res;
}

// replaces np.zeros()
function zeros(...dimensions) {
	const res = [];
	if (dimensions.length === 1) {
		while (dimensions[0]--) {
			res.push(0);
		}
	} else {
		while (dimensions[0]--) {
			res.push(zeros(...dimensions.slice(1)));
		}
	}
	return res;
}

class Layer_Dense {
	constructor(n_inputs, n_neurons) {
		this.weights = [];
		for (let i = 0; i < n_inputs; i++) {
			const temp = [];
			for (let j = 0; j < n_neurons; j++) {
				temp.push(2*Math.random() - 1);
			}
			this.weights.push(temp);
		}
		this.biases = zeros(n_neurons);
		this.output = null;
	}
	forward(inputs) {
		this.output = add(dot(inputs, this.weights), this.biases);
		return this;
	}
}

class Activation_ReLU {
	constructor() {
		this.output = null;
	}
	forward(inputs) {
		this.output = [];
		for (let i = 0; i < inputs.length; i++) {
			if (Array.isArray(inputs[i])) {
				this.output.push((new Activation_ReLU()).forward(inputs[i]).output);
			} else {
				this.output.push(Math.max(0, inputs[i]));
			}
		}
		return this;
	}
}

// Based on the code written by @vancegillies & updated by @daniel-kukiela
function spiral_data(points, classes) {
	const X = zeros(points * classes, 2);
	const y = Uint8Array.from(zeros(points*classes));
	const r_increment = 1 / (points - 1);
	const t_increment = 4 / (points - 1);
	for (let class_number = 0, i = 0; class_number < classes; class_number++) {
		let r = 0;
		let t = class_number * 4
		while (r <= 1 && t <= 4*(class_number + 1)) {
			const random_t = t * Math.random()*points * 0.2;
			X[i][0] = r * Math.sin(random_t*2.5);
			X[i][1] = r * Math.cos(random_t*2.5);
			y[i] = class_number;
			r += r_increment;
			t += t_increment;
			i++;
		}
	}
	return [X, y];
}

/*
const inputs = [0, 2, -1, 3.3, -2.7, 1.1, 2.2, -100];
const outputs = [];

for (let i = 0; i < inputs.length; i++) {
	//if (i > 0) {
		//outputs.push(inputs[i]);
	//} else {
		//outputs.push(0);
	//}
	outputs.push(Math.max(0, inputs[i]));
}

console.log(outputs);
*/

const [ X, y ] = spiral_data(100, 3);

const layer1 = new Layer_Dense(2, 5);
const activation1 = new Activation_ReLU();

layer1.forward(X);
//console.log(layer1.output);
activation1.forward(layer1.output);
console.log(activation1.output);

File: C#/p005-ReLU-Activation.cs
/*
 * Class Definition of ReLU Activation and spiral dataset
 * Based on nnfs.io by Harrison Kinsley & Daniel Kukiela
 * p005 from YT tutorial for the book: https://youtu.be/gmjzbpSVY1A
 * spiral dataset code based on: https://gist.github.com/Sentdex/454cb20ec5acf0e76ee8ab8448e6266c
 * spiral datase here defined as a function that returns a matrix X
 * 
 * Uses MathNet Numerics v4.11 by C.Ruegg, M. Cuda and J. Van Gael for vectors and matrix functions. 
 * To install in MSVS select Menu Project/Mangage NuGet packages...
 * in Browse tab search for MathNet.Numerics and install core package
*/
using System;
using MathNet.Numerics;
using MathNet.Numerics.LinearAlgebra;

namespace NNFS_p005
{
    class p005
    {
        static void Main(string[] args)
        {
            //in this case, spiral_data is defined as a function within the p005 class, main program.
            //therefore, to access it, it is necessary to first create an instance of p005
            p005 program_p005 = new p005();

            //Inputs, the matrix X is created using spiral_data which is a member of the object program_p005
            Matrix<double> X = program_p005.spiral_data(100, 3);

            //creates one dense layer and one activation layer:
            Layer_Dense layer1 = new Layer_Dense(2, 5); //2 inputs, 5 neurons
            Activation_ReLU activation1 = new Activation_ReLU();

            //call forward function of layer1 using X matrix as inputs
            layer1.Forward(X);

            //call forward function activation1 using outputs from layer 1
            activation1.Forward(layer1.output);

            //print results:
            //Console.WriteLine("\nInput matrix X (spiral data):");
            //Console.WriteLine(X.ToString());

            Console.WriteLine("Forward method layer 1, outputs1:");
            Console.WriteLine(layer1.output.ToString());

            Console.WriteLine("Forward method activation 1 ReLU, outputs1: (100x3) x 5 neurons");
            Console.WriteLine(activation1.output.ToString());

            Console.ReadLine();
        }

        //define dataset
        Matrix<double> spiral_data(int points, int classes)
        {
            //Matrix<double> X;
            //Vector<double> y;
            var M = Matrix<double>.Build; //shortcut to Matrix builder
            var V = Vector<double>.Build; //shortcut to Vector builder

            //build vectors of size points*classesx1 for y, r and theta
            var y = V.Dense(points * classes); //at this point this is full of zeros
            for (int j = 0; j < classes; j++)
            {
                var y_step = V.DenseOfArray(Generate.Step(points * classes, 1, (j + 1) * points));
                y = y + y_step;
            }
            var r = V.DenseOfArray(Generate.Sawtooth(points * classes, points, 0, 1));
            var theta = 4 * (r + y) + (V.DenseOfArray(Generate.Standard(points * classes)) * 0.2);
            var sin_theta = theta.PointwiseSin();
            var cos_theta = theta.PointwiseCos();

            var X = M.DenseOfColumnVectors(r.PointwiseMultiply(sin_theta), r.PointwiseMultiply(cos_theta));
            return X;
        }
    }

    //Define Layer_Dense class
    public class Layer_Dense
    {
        //attributes
        public Matrix<double> weithgts; //make them public attributs to be visible at program level
        public Matrix<double> biases;   //define as "horizontal" vector or 1xn matrix
        public Matrix<double> output;   //matrix with outputs

        //constructor
        public Layer_Dense(int n_inputs, int n_neurons)
        {
            //creats shortcuts to Matrix and Vector builders: 
            var M = Matrix<double>.Build;
            var V = Vector<double>.Build;

            //creates a n_inputs x n_neurons random matrix and assign to weights
            weithgts = M.Random(n_inputs, n_neurons);

            //creates a zero filled vector, this has to be a horizontal vector, hense using a 1xn matrix
            biases = M.Dense(1, n_neurons);
        }

        //Forward function
        public void Forward(Matrix<double> inputs)
        {
            //creates a Matrix of size: (batches x 1) , this is just an aux "vector"
            var M = Matrix<double>.Build;
            var v = M.Dense(inputs.RowCount, 1, 1);
            //multiply matrix v*biases: (batches x 1) dot (1 x neurons) = (batches x neurons) where each row is the same as the horizontal biases vector
            var biasm = v * biases; //biasm is a matrix where each row is identical and the rows are the biases horizontal vector
            //now bias matrix can be added to inputs*weights
            output = inputs * weithgts + biasm;
        }
    }

    //Define ReLU Activation Class
    public class Activation_ReLU
    {
        //attributes
        public Matrix<double> output;   //matrix with outputs

        //Forward function
        public void Forward(Matrix<double> inputs)
        {
            output = inputs.PointwiseMaximum(0);
        }
    }
}
